### ISSUE TITLE
FilePath.exists gives results inconsistent with FilePath.isdir

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2896]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2896|
|Type|defect|
|Created|2007-11-13 21:50:20Z|
|Branch|https://github.com/twisted/twisted/tree/filepath-exists-2896|

Consider this:

```
>>> from twisted.python.filepath import FilePath
>>> fp = FilePath('foobarbaz')
>>> fp.exists() # important, causes FilePath to cache a does-not-exist state
False
>>> fp.makedirs()
>>> fp.exists()
False
>>> fp.isdir()
True
```

<details><summary>Searchable metadata</summary>

```
trac-id__2896 2896
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_filepath_exists_2896 branches/filepath-exists-2896
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1194990620000000 1194990620000000
changetime__1203328474000000 1203328474000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [22487]) Branching to 'filepath-exists-2896'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [22488]) Force restat in exists, add a test.

Refs [#2896](https://github.com/twisted/twisted/issues/2896)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

`exists` docstring reads a bit awkwardly, _... can be stated correctly ..._, since "stated" is a different english word.

Typo in `test_crossMountMoveTo` docstring - _os.renamed_

The fix looks good, so far as it goes.  I wonder if `FilePath` wouldn't be better off just discarding the stat information after each mutating operation (and letting the next thing that cares do the restat - if it ever happens at all).  That's beyond the scope of this ticket though, I guess.

Feel free to merge once the docstrings are fixed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set status to `closed`|
|-|-|

(In [22593]) Merge filepath-exists-2896

Author: therve
Reviewer: exarkun
Fixes [#2896](https://github.com/twisted/twisted/issues/2896)

Force twisted.python.filepath.FilePath.exists to call restat: this prevents it
from using wrong cached value.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
